### DISCUSSION TITLE
大佬，我想学习下是如何实现下载测速的呢

### DISCUSSION THREAD

**Discussion by lei-9** [XIU2/CloudflareSpeedTest#325] [讨论]
我想的顺序：
第一步 根据cf 提供的可用网段 ipv4 / ipv6 来获取可用的网络ip（已实现）
第二步 随机ping 网络ip 是否可用（已实现）
第三步 测试ping 通的网络ip 下载速度（这一步需要怎么实现呢？求大佬赐教）
还是用 GET：{网络ip}/__down?bytes=100000  |  POST：{网络ip}/__up 测试上传下载吗？请求头一定要设置全吗？
原谅我看不懂go语言。。。求知若渴的小萌新诚恳求教～

**Comment by XIU2** [comment]
。。。不理解你到底要问什么
****
目前 CloudflareST 的测速流程是：
1. 解析 IP 段，随机生成一组 IP 用于测速
2. TCPing 延迟测速这些 IP
3. 将延迟测速结果排序（按延迟从低到高 + 丢包率从低到高）
4. 根据排序结果，从低到高的开始下载测速
5. 将下载测速结果排序
6. 输出结果
****
至于下载测速原理，就是指定 IP 的方式访问下载测速地址，根据下载总量除以下载时间得到平均速度。

其实和你在 Hosts 中指定测速域名的 IP，然后浏览器打开下载测速地址去下载是一个意思。
以及和 Curl 的 ` --resolve ` 参数实现的原理与效果一样。
```
curl -o /dev/null --resolve "测速地址的域名:443:IP" https://下载测速地址

# 示例：
curl -o /dev/null  --resolve "speed.cloudflare.com:443:172.67.172.187" https://speed.cloudflare.com/__down?bytes=200000000
```
即不再 DNS 解析域名，而是直接指定 IP 去访问该下载地址。

**Comment by lei-9** [comment]
> 

额，我的问题就是如何实现下载测速的，我按照你的方法来试试。感谢大佬解惑～
